### PR TITLE
Fixed reported errors for variadic element mismatches

### PR DESCRIPTION
**Pull request by Andarist**
fixes https://github.com/microsoft/TypeScript/issues/58697

### CODE REVIEW

**Review by jakebailey — APPROVED** [review]
This seems correct to me? @weswigham

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -23411,7 +23411,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                     const sourceArity = getTypeReferenceArity(source);
                     const targetArity = getTypeReferenceArity(target);
                     const sourceRestFlag = isTupleType(source) ? source.target.combinedFlags & ElementFlags.Rest : ElementFlags.Rest;
-                    const targetRestFlag = target.target.combinedFlags & ElementFlags.Rest;
+                    const targetHasRestElement = target.target.hasRestElement;
```
Comment (weswigham) [inline src/compiler/checker.ts:23414]: `type.hasRestElement = !!(combinedFlags & ElementFlags.Variable);`, and `ElementFlags.Variable` is `ElementFlags.Rest | ElementFlags.Variadic` so... yeah, I guess this works to expand the check from just rest to rest and variadic elements! I read this like 3 times before checking how `hasRestElement` was defined and realizing that, though.

We should slap a `@deprecated` on `hasRestElement`, really.  The element flags are really the new way to check that.

**Review by weswigham — APPROVED** [review]

**Comment by weswigham** [comment]
Looks like this one needs a small resync with `main` before it can be merged, though~

**Comment by Andarist** [comment]
I synced this with main, it should now be ready to merge :)

File: src/compiler/checker.ts (hunk 1 above)
Comment (Andarist) [reply src/compiler/checker.ts]: > We should slap a @deprecated on hasRestElement, really. The element flags are really the new way to check that.

https://github.com/microsoft/TypeScript/pull/58864 😉

**Pull request merged**